Source/WebCore/rendering/MotionPath.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "RenderLayerModelObject.h"
+
+namespace WebCore {
+
+class RenderElement;
+class PathOperation;
+class RayPathOperation;
+
+struct MotionPathData {
+    FloatRect containingBlockBoundingRect;
+    FloatPoint rayStartingPosition;
+};
+
+class MotionPath {
+public:
+    static std::optional<MotionPathData> populateMotionPathDataForRenderer(const RenderElement&);
+    static bool needsUpdateAfterContainingBlockLayout(const PathOperation&);
+    static void applyMotionPathTransform(const RenderStyle&, const TransformOperationData&, TransformationMatrix&);
+    static std::optional<Path> computePathForRay(const RayPathOperation&, const TransformOperationData&);
+    static double lengthForRayPath(const RayPathOperation&, const MotionPathData&);
+    static double lengthForRayContainPath(const FloatRect& elementRect, double computedPathLength);
+};
+
+class TransformOperationData {
+public:
+    TransformOperationData(FloatRect boundingBox)
+        : m_boundingBox(boundingBox)
+    {
+    }
+
+    TransformOperationData(FloatRect boundingBox, const RenderElement* renderer)
+        : m_boundingBox(boundingBox)
+    {
+        if (renderer)
+            m_containingBlockData = MotionPath::populateMotionPathDataForRenderer(*renderer);
```

Comment:
Feels a bit odd this named m_containingBlockData when the struct is named MotionPathData

Reply:
Changed to m_motionPathData.